src/util/gamma_picker.cpp

```diff
@@ -59,8 +61,11 @@ namespace lws
       const std::size_t outputs_to_consider = rct_offsets.back() - initial;
 
       static_assert(0 < DIFFICULTY_TARGET_V2, "block target time cannot be zero");
+      // match wallet2's integer truncation implementation to avoid creating distinct anonymity puddles
       // this assumes constant target over the whole rct range
-      outputs_per_second = outputs_to_consider / double(DIFFICULTY_TARGET_V2 * blocks_to_consider);
+      seconds_per_output = DIFFICULTY_TARGET_V2 * blocks_to_consider / outputs_to_consider;
```

Comment:
Please flip back to the original logic (`outputs_per_second`), this removes a division in the `operator()` function.

Reply:
The point of this is to match integer truncation in wallet2 (https://github.com/monero-project/monero/pull/7845)

Reply:
There doesn't seem to be any justification for integer truncation in that post. Just you claiming that two other PRs were more important. If anything, `wallet2` should be moving towards this implementation and not the other way.

Reply:
> Also, I'm still serious about flipping that division logic. Not only is division more costly (barely), but seconds_per_output is less natural than outputs_per_second (at least I think so).

I much prefer the `seconds_per_output` approach as well, for both reasons too.

My comment explains that it's not necessarily a clear decision whether it is safe to patch integer truncation. I have (and had) more analysis that shows why it's not a clear decision -- I will share it here if you feel it's relevant. Continuing to pile onto that issue seemed to be overly confusing.

I agree that `wallet2` should ideally move towards removing integer truncation (and move toward this `outputs_per_second` approach), which is what I'm *also* continuing to work on fully assessing. Can just hold off on this PR until that work is complete if you'd like.

Reply:
> I have (and had) more analysis that shows why it's not a clear decision -- I will share it here if you feel it's relevant. Continuing to pile onto that issue seemed to be overly confusing.

The primary argument I've seen is that you can't force an upgrade until the next hardfork  (uniformity issue). Is there another? Because I think you are making the integer truncation change too complicated -

MyMonero/Monero-LWS (the former representing a decent percentage of the transactions), didn't have the integer truncation issue (or the incorrect curve fit), so there's already a decent number of users with both changes. Which means there's already a lack of uniformity in the integer truncation case. And there's low probability of a benefit to an unintended truncation (it was unintended and accidentally made it better?).

Reply:
I'm not seeing a compelling reason to accept the integer truncation issue (at least not yet). I have several questions, and possible criticisms of this analysis, which I am moving to the referenced Github issue.

This implementation was/is `wallet2` users, and now you want me to _under select_ new outputs, when you've done no analysis on **real** (as opposed to decoy) outputs.

Reply:
> This implementation was/is protecting wallet2 users, and now you want me to under select new outputs

I think it's definitely reasonable to not accept on these grounds. I figured it was desirable to have uniform implementations across wallet2 users and non-wallet2 users to ensure tx uniformity, but hearing your position I would agree that isn't absolutely critical or a strong reason enough in and of its own, without further evidence, to move away from integer truncation here.

> when you've done no analysis on real (as opposed to decoy) outputs.

The blue spikes = analysis on real outputs, but the crux of the position (to leave integer truncation unpatched) isn't well defended by those blue spikes.

When I said:

> I agree that wallet2 should ideally move towards removing integer truncation (and move toward this outputs_per_second approach), which is what I'm also continuing to work on fully assessing. Can just hold off on this PR until that work is complete if you'd like.

I will reword:

> I agree, the safer thing to do would be to just keep this PR on hold until further analysis on **real** outputs is complete.

This is what I intended to spend time working on.

Reply:
FWIW, I think the following chart from [this comment](https://github.com/monero-project/research-lab/issues/86#issuecomment-915806414):

![mymonero recent shift from 2 to 1](https://user-images.githubusercontent.com/26468430/132763632-407f5fa9-58de-4797-acc8-c5e281a49f05.png)

is the visualized reason *not* to introduce integer truncation into monero-lws without more compelling evidence. Evidently this PR as is would introduce a larger under-selection for recent outputs (since green dash is below blue - - and red line is above blue -- for most of the range). And the cost of introducing that under-selection should be weighed against the benefit of higher tx uniformity we get from consistent implementations across wallet2/non-wallet2 users.

In hindsight, I shouldn't have tried to bring truncation to monero-lws without stronger reasoning. Your critical eye is appreciated.

Reply:
@vtnerd I just want to chime in here and give some context. If all goes according to plan, this integer truncation issue will be moot in a few months since hopefully we will be moving to a brand new parametric distribution for the mixin selection algorithm.

The original suggestion of [Moser et al. (2018)](https://www.sciendo.com/article/10.1515/popets-2018-0025), which the current code attempts to implement, was a decent first draft, but in my view it never should have been implemented in production code in 2018, much less persisted until 2021. I base this view on my extensive training and experience in statistical analysis as an empirical microeconomist as well as Monero-specific theoretical and empirical investigations that I, @j-berman , and others have carried out recently. I am developing a CCS proposal to fund the research required to determine the best way forward in overhauling the mixin selection algorithm.

I am not sure about the merits of going with one fix or another with respect to the integer truncation issue. All that I wanted to say is that any fix will likely be quite temporary.